compiler_gym/envs/llvm/datasets.py

```diff
@@ -110,12 +158,519 @@
+    _environ = os.environ.copy()
+    try:
+        yield
+    finally:
+        os.environ.clear()
+        os.environ.update(_environ)
+
+
+@contextmanager
+def benchmark_execution_environment(env: Dict[str, str]):
+    """Setup the os.environ state for executing benchmarks."""
+    with temporary_environment():
+        for var in os.environ:
+            if not var.startswith("COMPILER_GYM_SITE_DATA") and var not in {
+                "PATH",
+                "RUNFILES_DIR",
+                "RUNFILES_MANIFEST_FILE",
+                "USER",
+                "PWD",
+                "SHELL",
+                "TERM",
+                "TMPDIR",
+                "HOME",
+            }:
+                del os.environ[var]
+        for key, val in env.items():
+            os.environ[key] = expand_command_vars(val)
+        yield
+
+
+def expand_command_vars(cmd: str) -> str:
+    """Expand shell variables in a command."""
+    with temporary_environment():
+        os.environ.clear()
+        os.environ["BIN"] = "./a.out"
+        os.environ["D"] = str(_CBENCH_DATA)
+        return os.path.expandvars(cmd)
+
+
+_VALIDATORS: Dict[str, List[Callable[["LlvmEnv"], Optional[str]]]] = defaultdict(list)
+
+
+def validator(
+    benchmark: str,
+    cmd: str,
+    data: Optional[List[str]] = None,
+    outs: Optional[List[str]] = None,
+    platforms: Optional[List[str]] = None,
+    compare_output: bool = True,
+    validate_result: Optional[
+        Callable[[BenchmarkExecutionResult], Optional[str]]
+    ] = None,
+    linkopts: List[str] = None,
+    env: Dict[str, str] = None,
+    pre_execution_callback: Optional[Callable[[], None]] = None,
+) -> bool:
+    """Declare a new benchmark validator.
+
+    :param benchmark: The name of the benchmark that this validator supports.
+    :cmd: The shell command to run the validation. Variable substitution is applied to this value as follows: :code:`$BIN` is replaced by the path of the compiled binary and :code:`$D` is replaced with the path to the benchmark's runtime data directory.
+    :data: A list of paths to input files.
+    :outs: A list of paths to output files.
+    :return: :code:`True` if the new validator was registered, else :code:`False`.
+    """
+    platforms = platforms or ["linux", "macos"]
+    if {"darwin": "macos"}.get(sys.platform, sys.platform) not in platforms:
+        return False
+    infiles = [_CBENCH_DATA / p for p in data or []]
+    outfiles = [Path(p) for p in outs or []]
+    linkopts = tuple(linkopts or [])
+    env = env or {}
+
+    _VALIDATORS[benchmark].append(
+        make_cBench_validator(
+            cmd=cmd,
+            input_files=infiles,
+            output_files=outfiles,
+            compare_output=compare_output,
+            validate_result=validate_result,
+            linkopts=linkopts,
+            os_env=env,
+            pre_execution_callback=pre_execution_callback,
+        )
+    )
+    return True
+
+
+def _compose_validators(*validators):
```

Comment:
This seems not to be called?

Reply:
Called in `get_llvm_benchmark_validation_callback()`, but it wasn't clear so I've refactored.